cartography/intel/okta/groups.py

```diff
@@ -1,298 +1,390 @@
+    _cleanup_okta_groups(neo4j_session, common_job_parameters)
+
+    logger.info("Syncing Okta group rules")
+    group_rules = asyncio.run(_get_okta_group_rules(okta_client))
+    transformed_group_rules = _transform_okta_group_rules(group_rules)
+    _load_okta_group_rules(
+        neo4j_session, transformed_group_rules, common_job_parameters
+    )
+    _cleanup_okta_group_rules(neo4j_session, common_job_parameters)
 
-        paged_results = PagedResults(paged_response, UserGroup)
 
-        group_list.extend(paged_results.result)
+@timeit
+async def _get_okta_groups(okta_client: OktaClient) -> List[OktaGroup]:
+    """
+    Get Okta groups list from Okta
+    :param okta_client: An Okta client object
+    :return: List of Okta groups
+    """
+    output_groups = []
+    query_parameters = {"limit": 200}
+    groups, resp, _ = await okta_client.list_groups(query_parameters)
+    output_groups += groups
+    while resp.has_next():
+        groups, _ = await resp.next()
+        output_groups += groups
+        logger.info(f"Fetched {len(groups)} groups")
+    return output_groups
 
-        check_rate_limit(paged_response)
 
-        if not is_last_page(paged_response):
-            next_url = paged_response.links.get("next").get("url")
-        else:
-            break
-
-    return group_list
+@timeit
+def _transform_okta_groups(
+    okta_client: OktaClient,
+    okta_groups: List[OktaGroup],
+    okta_group_roles: List[OktaGroupRole],
+) -> List[Dict[str, Any]]:
+    """
+    Convert a list of Okta groups into a format for Neo4j
+    :param okta_client: An Okta client object
+    :param okta_groups: List of Okta groups
+    :param okta_group_roles: List of Okta group roles
+    :return: List of group dicts
+    """
+    transformed_groups: List[Dict] = []
+    logger.info(f"Transforming {len(okta_groups)} Okta groups")
+    for okta_group in okta_groups:
+        group_props = {}
+        group_props["id"] = okta_group.id
+        group_props["created"] = okta_group.created
+        group_props["last_membership_updated"] = okta_group.last_membership_updated
+        group_props["last_updated"] = okta_group.last_updated
+        group_props["object_class"] = json.dumps(okta_group.object_class)
+        group_props["profile_description"] = okta_group.profile.description
+        group_props["profile_name"] = okta_group.profile.name
+        group_props["group_type"] = okta_group.type.value
+        # For each group, grab what users might assigned
+        group_members: List[Dict] = asyncio.run(
+            _get_okta_group_members(okta_client, okta_group.id)
+        )
+        for group_member in group_members:
+            match_user = {**group_props, "user_id": group_member.id}
+            transformed_groups.append(match_user)
+        # Check to see if this group has any matching group roles
+        # TODO: Converting this into a hashmap would make perform better
```

Comment:
This feels like _a lot_ of wasted work since the group-role list can be very big. Let's take some time to think about restructuring.